feat(spectator): added expectConcurrent to spectator-http (#231)

dedwardstech opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/spectator/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?
An addition to the SpectatorHttp API.
<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
`spectator.expectOne()` calls `this.controller.verify()` which causes tests to fail if there are concurrent requests going out.
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: 231

## What is the new behavior?
Adding `spectator.expectConcurrent()` allows for expecting multiple requests.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

dedwardstech commented:
A little rational behind adding this function.

Adding a flag to `expectOne` felt bulky and after a string of `expectOne(url, method, flag)`, you would then have to manually call `spectator.controller.verify()`. By adding this function you declare your expectations and the function will call verify for you, just like how `expectOne` works now.

NetanelBasal commented on README.md:
```diff
@@ -786,6 +786,49 @@ We need to create an HTTP factory by using the `createHttpFactory()` function, p
+        { URL: '/api2/todos', method: HttpMethod.GET }
+    ]);
+
+    reqs.forEach(req => {
```
I like this approach, but I think we can make it cleaner. Something like: `flushAll()`

dedwardstech replied:
Are you talking about exposing another method on SpectatorHttp?

Possibly something like `spectator.flushAll(requests, [ { foo: 'lorem' }, { bar: 'ipsum' }])`?

NetanelBasal replied:
Yes

dedwardstech replied:
Okay, I'll make those changes.

NetanelBasal commented on projects/spectator/test/todos-data.service.spec.ts:
```diff
@@ -35,6 +35,21 @@ describe('HttpClient testing', () => {
+    const spectatorHttp = http();
+
+    spectatorHttp.service.concurrentRequests().subscribe();
+    const reqs = spectatorHttp.expectConcurrent([
```
It can be cleaner: 
`spectatorHttp.expectConcurrent(['one', 'two', 'three'].map(url => 
    ({ url, method:HttpMethod.POST})))`

NetanelBasal commented on README.md:
```diff
@@ -786,6 +786,49 @@ We need to create an HTTP factory by using the `createHttpFactory()` function, p
 - `get()` - A proxy for Angular `TestBed.get()`
 - `expectOne()` - Expect that a single request was made which matches the given URL and it's method, and return its mock request
 
+Spectator also makes testing concurrent requests easier. Similar to `spectator.expectOne()`, you can use,
```
Thanks for the effort, but I prefer to combine this example with the previous one. Please add the relevant code to previous snippet, and leave the method and the description here. Thanks.

dedwardstech commented:
I've added the flush all method, cleaned up the test for it and modified the documentation. For the documentation, I just merged the two together because it seemed to flow better.

I've given some thought to storing the pending requests for `expect(One/Concurrent)` in SpectatorHttp. It seems like it would make everything flow pretty nicely. Given that spectator is something of a convenience wrapper around the test bed, it would make sense to be able to call `spectator.flush()` or `spectator.flushAll()` instead of storing the requests in a variable and then flushing.

State could be stored pretty painlessly in spectator, but doing so might cause a breaking change. This is just something I thought about while working on this change. I'd be happy to put in a PR if this makes sense and there aren't things I'm missing here.
